### ISSUE TITLE
[diffusion] update_weights_from_disk transformer reload fails (shape mismatch 400) — diffusers checkpoint layout not reconciled with sglang params

### ISSUE THREAD

**Issue by kangwangamd** [sgl-project/sglang#31924]
### Summary

`multimodal_gen` `update_weights_from_disk` fails when updating the `transformer` module of diffusion models loaded from a diffusers-format checkpoint. The request returns **HTTP 400** with an empty error message; the server-side cause is a bare `AssertionError` from a weight `shape` mismatch. Reproduced in CI by `test_update_weights_from_disk.py` for **both** `Qwen-Image` and `FLUX.2-klein-base-4B`.

> Note: this is **not** the InferenceMode-500 issue fixed by #31263 (a separate failure in the same test file). This is a distinct, still-open failure.

### Repro (CI)

`multimodal-gen-test-1-gpu-amd` shard 3 (standalone partition), e.g. run 29743214297 / job 88354737021. Failing cases: `test_update_weights_from_disk_default`, `test_update_weights_specific_modules`, `test_corrupted_weights_rollback`, `TestUpdateWeightsFromDiskWithOffload::test_update_weights_with_offload_enabled`.

Server log + traceback:

```
Flux2Transformer2DModel: Skipping weight update: parameter 'img_in.bias'
    not found after mapping to 'x_embedder.bias'
Weight update failed for module 'transformer': .
POST /update_weights_from_disk  400 Bad Request

  weights_updater.py:295 load_weights_into_model -> weight_loader(param, loaded_weight...)
  linear.py:410 weight_loader
      assert param_data.shape == loaded_weight.shape
  AssertionError
```

### Root cause

The reload path (`weights_updater.load_weights_into_model`) applies the model's `param_names_mapping` to translate diffusers checkpoint keys (`img_in`, `txt_in`, ...) into sglang parameter names (`x_embedder`, `context_embedder`, ...), then copies each tensor via the parameter's `weight_loader`. For at least one `transformer` parameter the mapped tensor reaches `ColumnParallelLinear.weight_loader` (plain 2-arg loader, TP=1) and fails `assert param_data.shape == loaded_weight.shape` — i.e. the diffusers on-disk layout/shape differs from the sglang parameter, and the reload path does not apply the same per-parameter reshape that the **initial** checkpoint load applies. The assertion is bare, hence the empty-message 400.

Secondary: the `FSDPQwen3ForCausalLM` text encoder produces thousands of `not found after mapping to '<same name>'` skips — the mapping doesn't cover the FSDP-flattened parameter names, so the text-encoder reload is effectively a no-op.

### Not the cause (checked)

A fused-QKV-shard hypothesis (reload dropping `merge_index`/`total_split_params`) was tested against the real FLUX.2-klein-base-4B checkpoint and ruled out: the FLUX `param_names_mapping` has **0** merge entries, **0/169** disk tensors get a `merge_index`, and QKV is only fused for `ModelOptFp4Config` packed checkpoints (the BF16 test checkpoint builds separate `to_q/to_k/to_v` that already match disk). So this is a plain layout/shape reconciliation gap, not a fusion issue.

### CI coverage

This standalone test runs on the AMD lane only; the NVIDIA diffusion partition plan sets `standalone_files: []`, so NVIDIA does not exercise `update_weights_from_disk`. The defect is device-independent (weight-name/shape logic); it simply surfaces only on AMD.

### Interim CI action

Because this fails deterministically and blocks the AMD `multimodal-gen-test-1-gpu` shard every run — while NVIDIA already does not run it — a companion PR skips `test_update_weights_from_disk` on ROCm (via the existing `current_platform.is_hip()` guard in `gpu_cases.py`) to unblock the lane. The est-time entry is left in place so re-enabling after a fix is a one-line revert. This issue tracks the real fix.

### Asks

- Pin the exact failing parameter (`name`, `param.shape` vs `loaded_weight.shape`) by booting the real transformer / running the test with logging at the mismatch point.
- Reconcile the diffusers->sglang layout for that parameter — either extend `param_names_mapping` (rename + reshape) or apply the initial loader's per-parameter reshape in the reload path.
- Separately, cover the FSDP text-encoder name mapping (or document why it is skipped), then remove the ROCm skip.

cc diffusion / multimodal_gen post-training maintainers.

**Comment by firefly-silvers** [comment]
I spent some time on this and ended up with results that differ from the issue description in a few places, so posting the measurements here. #33588 fixes part of it.

## The headline symptom does not reproduce on NVIDIA

All 5 tests in `test_update_weights_from_disk.py` **pass** on 1×NVIDIA L4 with `FLUX.2-klein-base-4B`, both modes including `--dit-layerwise-offload`.

The transformer had **zero** skipped weights.

## Why the transformer passes here — the test never exercises the mapping

`FLUX.2-klein-base-4B/transformer/` holds 169 tensors already named exactly like sglang's parameters:

```
x_embedder.weight
context_embedder.weight
norm_out.linear.weight
single_transformer_blocks.0.attn.to_qkv_mlp_proj.weight
```

So every name matches at the `if name in model_params` short-circuit in `_iter_module_weight_updates`, and `param_names_mapping` never runs at all. The reload path for the DiT is effectively an identity copy in this test — on any platform.

The repo also ships a second, BFL-format copy of the transformer at the repo root(`flux-2-klein-base-4b.safetensors`, 149 tensors: `double_blocks`, `img_attn.qkv`, `.scale`, `img_in.weight`). The reload path only ever reads `<model>/transformer/`, so it never sees that file. For what it is worth, **neither** file contains `img_in.bias` — the key named in the log in the issue description — so I could not track down where that came from.

## What does reproduce, every run: silent weight drops

Across the same 5 tests, **2409 tensors were discarded** and every request still returned `success: true`:

| module | dropped |
|---|---|
| `Qwen3ForCausalLM` (text encoder) | 1990 |
| `FSDPQwen3ForCausalLM` (text encoder, offload class) | 398 |
| `AutoencoderKLFlux2` (vae) | 21 |
| `Flux2Transformer2DModel` | 0 |

The tests stayed green because they only assert the transformer checksum — the file's own NOTE (lines 29-40) explains that vae/text_encoder checksums were deliberately not compared.

## Correcting the secondary diagnosis

The issue says the text-encoder problem is that "the mapping doesn't cover the FSDP-flattened parameter names." Measured on `FSDPQwen3ForCausalLM` itself:

| | non-offload `Qwen3ForCausalLM` | `FSDPQwen3ForCausalLM` |
|---|---|---|
| parameter count | 290 | 290 |
| sample names | `layers.0.self_attn.qkv_proj.weight` | identical |
| skips per update | 398 | 398 |

Identical names and count — nothing is flattened, in either class. The actual causes are:

1. **A `model.` prefix.** Disk has `model.layers.7...`; parameters are `layers.7...`. Startup strips it (`encoders/qwen3.py:431`); the reload path does not, so **all** 398 tensors miss - including unfused ones like `o_proj` and the layernorms.
2. **A missing `shard_id`.** `weights_updater.py` calls `weight_loader(param, weight)` with two arguments; startup passes three (`encoders/qwen3.py:467`). Without it, `MergedColumnParallelLinear.weight_loader` assumes the tensor is already fused on disk.

Arithmetic check: 398 disk tensors − (36 layers × 3 fused) = 290 = the observed parameter count.

## Also present, not previously reported: VAE buffers

3 BatchNorm buffers (`running_mean`, `running_var`, `num_batches_tracked`) were never reloaded, because the reload path builds its lookup from `named_parameters()`. For this VAE the BatchNorm is `affine=False`, so those buffers are the layer's entire state — and they set the scale/shift applied to every decode (`configs/pipeline_configs/flux.py:703-717`).

They were invisible to every existing check: `iter_materialized_weights` yields `named_parameters()` only, so no checksum in the suite can see a buffer.

## What #33588 does and does not do

Fixes the text encoder, the VAE buffers, and adds per-module loaded counts to the response - **2409 → 398** dropped weights, tests still passing.

Still open, and marked with a `TODO` in the code:

- **FSDP-sharded text encoders** (the remaining 398). `load_weights` runs before `shard_model()` at startup, so it has no DTensor branch; calling it on a wrapped module raises `aten.copy_.default got mixed torch.Tensor and DTensor`. Excluded by an explicit guard for now. No longer silent — it surfaces as `text_encoder: 0`.
- **The headline transformer failure.** Untouched, so this does **not** unblock re-enabling the ROCm skip from #31925.

## Headline still pending root cause

I'm trying to get access to an AMD box to chase this. In the meantime, if someone already has one, the two most useful things would be: 
1. whether Flux2Transformer2DModel skips anything there — here it skips nothing, which suggests the AMD run saw different checkpoint contents;
2. the lines just before the AssertionError, with the parameter name and both shapes. The assert at linear.py:421 is bare, which is why the 400 came back empty.
